### ISSUE TITLE
Ensure drush works for Drupal 8.4

### ISSUE THREAD

**Issue by beeradb** [drud/ddev#463]
## What happened (or feature request):
As @weitzman pointed out in #461:

> Among other goodies, this gets you a site local drush which is the only real supported way to run drush with Drupal 8.4 (coming next month).

Currently, we're using a global Drush install in ddev the nginx-php container, so it's pretty likely we'll be broken. 

## What you expected to happen:
We'll need to test against the upcoming 8.4 release and ensure we can still support drush from within the nginx container. 

I'd recommend we switch our Drupal8 tests away from the `drud/drupal8` repo, and instead have them use the composer based install method. Additionally, it would be great to add some automation coverage via `ddev exec` to ensure Drush is working as expected.

**Comment by weitzman** [comment]
A recent version of Drush8 as a global Drush actually works properly - it hands of to the site local drush.

However, to the extent people want a global Drush, the [drush-launcher](https://github.com/drush-ops/drush-launcher) is now recommended. Its a separate program from Drush.

**Comment by rickmanelius** [comment]
I added myself as an assignee and pulled to our planning board.

**Comment by rickmanelius** [comment]
Given that Drupal 8.4 is going to release is going to be next week (https://groups.drupal.org/node/517148), it would probably make sense to prioritize this. It sounds like we need the following:

* [ ] Change the tests from our repo snapshot to the official D8 repository (note this will add testing time).
* [ ] Resolve any issues related to drush, notably the installation of the correct/updated versions within the ngninx container and whether or not drush-launcher adds value.

Sidtenoe: I'm not entirely sure the extent by which drush-launcher would provide value, but I haven't tested the composer based build on 8.4.x-rc2. Given that each site has their own container, I'd guess the scenario that it would apply is if each site specified its own version of drush and running `ddev exec drush` would require drush-launcher to ensure that the installed version was the targetted one. Is that correct?

**Comment by rickmanelius** [comment]
If the above comment is valid, I'd consider this actionable able and we can prioritize in the upcoming milestone.

**Comment by weitzman** [comment]
Yes thats correct.

**Comment by rickmanelius** [comment]
@rfay I believe this is actionable, but let me know if there is additional detail or criteria to stub out before you/@nmccrory are comfortable moving on it.

**Comment by rickmanelius** [comment]
Not sure if this is caused by changes in the Drupal 8.4.x codebase (requiring composer by default), but in attempting to review https://github.com/nmccrory/ddev/blob/180ba2ddafbcfb724ac8d2230ed39f833f4f4023/docs/users/cli-usage.md where we're building D8 from composer, when I was using the PR from @rfay introducing flags for `ddev config` (see https://github.com/drud/ddev/issues/476), a  standard `ddev config` was not executing properly. I'll need to do more digging, but I wanted to surface this issue.

**Comment by rfay** [comment]
Please mention exactly what kind of problems you had with `ddev config` - it's super important! Could you give a screenshot or something?

**Comment by tannerjfco** [comment]
Based on my testing, the version of drush we currently provide - 8.1.12 - hands off to a site local drush as expected for recent drush releases. I was able to provision a 8.4.0 codebase successfully from drupal-composer, and the drush version provided with the project was used unless i was completely outside the codebase. As things currently stand, I don't think we _need_ to take any action to support the 8.4 release.

Looking at drush-launcher, this seems like it would be fine for D8 projects setup to use composer, but anything else we wouldn't be providing a working drush setup for. That would mean any previous D7 or D8 codebase would need to be refactored to use composer and include a site-local drush in order to have a working drush setup. I don't think we should be imposing such a requirement on our users, as there's no telling whether such a task is reasonably achievable for any given site. Given this, I think we should stick with the global drush install for the time being. 

While a drush update doesn't seem strictly necessary at this time, I do have a PR to update it and a couple other tools in the container, along with some other maintenance.

**Comment by rickmanelius** [comment]
Hi @tannerjfco. I haven't reviewed the PR specifically, but given the scope of this issue is to ensure that Drupal 8.4 is functional, I think the plan of sticking with the smaller changes for now makes sense while we work through the implications on Drupal 7 / previous Drupal 8 as a separate issue.

One other note. One of the recommendations was to test against drupal/drupal8 instead of our drud/drupal8 snapshot. I assume this would also be an appropriate place to make that change so we can start tracking against all future releases.

**Comment by tannerjfco** [comment]
I think we're in good shape for D8.4, so I'm going to close this given the following:

- Our previous drush version 8.1.12 was sufficient to support Drupal 8.4
- We updated drush anyways to latest 8.1.15
- We have switched our tests to use the official D8 release instead of our test repo. We'll have to update our test archive URL periodically in order to test the latest release, but this is already a better situation than we had.

If we encounter any new issues related to 8.4, a new ticket can be filed at such time.
